### ISSUE TITLE
Tpm error after OS upgrade

### ISSUE THREAD

**Issue by nesvera** [tpm2-software/tpm2-tss#1927]
I've being struggling with a tpm error in the last days.

```
root@daniel:~# openssl req -new -engine libtpm2tss -keyform engine -key ~/test.pem -out ~/teste.csr -subj '/CN=blah/'
engine "tpm2tss" set.
WARNING:esys:../tpm2-tss-2.3.3/src/tss2-esys/api/Esys_Load.c:324:Esys_Load_Finish() Received TPM Error
ERROR:esys:../tpm2-tss-2.3.3/src/tss2-esys/api/Esys_Load.c:112:Esys_Load() Esys Finish ErrorCode (0x000001df)
140400698876480:error:8007206D:tpm2-tss-engine:init_tpm_key:Unknown TPM error occured. Please check tpm2tss logs:src/tpm2-tss-engine-common.c:566:
140400698876480:error:8008C06D:tpm2-tss-engine:rsa_priv_enc:Unknown TPM error occured. Please check tpm2tss logs:src/tpm2-tss-engine-rsa.c:160:
140400698876480:error:8007406F:tpm2-tss-engine:esys_ctx_free:Some unknown error occured:src/tpm2-tss-engine-common.c:106:
140400698876480:error:0D0DC006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-1.1.1g/crypto/asn1/a_sign.c:224:
```

This computer was running the version 1.0 of a custom OS when the private key was generated and stored on TPM.

```
tpm2tss-genkey -a rsa ~/test.pem
openssl req -new -engine libtpm2tss -keyform engine -key ~/test.pem -out ~/teste.csr -subj '/CN=blah/'
openssl rsa -engine libtpm2tss -inform engine -in ~/test.pem -pubout -outform pem -out ~/test.pub
tpm2_nvdefine 0xXXXXXXX -C 0xXXXXXXXX -s XXX -a 0xXXXXX
tpm2_nvwrite 0xXXXXXXX -C 0xXXXXXXXX -i ~/test.pem
```

After some point the computer was updated to a new version of the OS that has a different linux kernel, and then the error with tpm appeared. The private key persist during the upgrade. If I downgrade the OS again, the problem vanishes. Tpm_clear is not an option here because I need to continue using the value stored on it.

TPM2-tss on version 1.0 of the OS: v2.3.2
TPM2-tss on version 2.0 of the OS: v2.3.3

Does the tpm's commands check OS integrity for every command?

**Comment by AndreasFuchsSIT** [comment]
The integrity error refers to the HMAC over the TPM command parameters, not the integrity of the system itself.

In the step from 2.3.2 to 2.3.3 we did touch these areas according to changelog, but I though we fixed bugs. This is the first case of a regression I've seen.

Could you maybe try to upgrade to 2.4.5 ?

Also, which version of the tpm2-tss-engine are you using ? There's also been a new release with 1.1.0 ?

**Comment by nesvera** [comment]
Thank you for the fast reply. I will try to upgrade tpm2-tss to 2.4.5.

tpm2-tss-engine on version 1.0 of the OS: v1.0.0
tpm2-tss-engine on version 2.0 of the OS: v1.1.0-rc1

**Comment by nesvera** [comment]
I upgraded tpm2-tss to 2.4.5, and tpm2-tss-engine to the current 1.1.x. The error continues the same as before.

```
root@daniel:~# openssl req -new -engine libtpm2tss -keyform engine -key ~/test.pem -out ~/teste.csr -subj '/CN=blah/'
engine "tpm2tss" set.
WARNING:esys:../tpm2-tss-2.4.5/src/tss2-esys/api/Esys_Load.c:324:Esys_Load_Finish() Received TPM Error 
ERROR:esys:../tpm2-tss-2.4.5/src/tss2-esys/api/Esys_Load.c:112:Esys_Load() Esys Finish ErrorCode (0x000001df) 
139884002722368:error:8007206D:tpm2-tss-engine:init_tpm_key:Unknown TPM error occured. Please check tpm2tss logs:src/tpm2-tss-engine-common.c:570:
139884002722368:error:8008C06D:tpm2-tss-engine:rsa_priv_enc:Unknown TPM error occured. Please check tpm2tss logs:src/tpm2-tss-engine-rsa.c:160:
139884002722368:error:8007406F:tpm2-tss-engine:esys_ctx_free:Some unknown error occured:src/tpm2-tss-engine-common.c:106:
139884002722368:error:0D0DC006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-1.1.1g/crypto/asn1/a_sign.c:224:
```

You mentioned "The integrity error refers to the HMAC over the TPM command parameter". Is it possible that the new bios of the new OS is changing some parameter inside TPM?

**Comment by AndreasFuchsSIT** [comment]
No. that has nothing to do with the error code.

Each command sent to the TPM is wrapped in a session and HMAC'ed.
The error code means that this HMAC is false.

I fear that we introduced some bug with 2.3.3 that you now encounter, but I don't know which...

Could you git-bisect between 2.3.2 and 2.3.3 and give me a hint ?

**Comment by nesvera** [comment]
I made a mistake. I said before that the initial tpm2-tss version (OS v1.0) was v2.3.2, but instead it is v2.2.3 revision 36b1539c82bf675265d6f6a6cd808a189b6971f4.

I tried to downgrade it to the version v2.2.3 and the error disappeared, and I was able to run the openssl command.

Then, we can argument that the problem was introduced between v2.2.3 and v2.3.2.

I will try the git-bisect.

**Comment by nesvera** [comment]
I downgraded tpm2-tss to v2.3.0 and the problem continues to happen.

```
WARNING:esys:../tpm2-tss-2.3.0/src/tss2-esys/api/Esys_Load.c:320:Esys_Load_Finish() Received TPM Error 
ERROR:esys:../tpm2-tss-2.3.0/src/tss2-esys/api/Esys_Load.c:112:Esys_Load() Esys Finish ErrorCode (0x000001df) 
140592290517568:error:8007206D:tpm2-tss-engine:init_tpm_key:Unknown TPM error occured. Please check tpm2tss logs:src/tpm2-tss-engine-common.c:566:
140592290517568:error:8008C06D:tpm2-tss-engine:rsa_priv_enc:Unknown TPM error occured. Please check tpm2tss logs:src/tpm2-tss-engine-rsa.c:160:
140592290517568:error:8007406F:tpm2-tss-engine:esys_ctx_free:Some unknown error occured:src/tpm2-tss-engine-common.c:106:
140592290517568:error:0D0DC006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-1.1.1g/crypto/asn1/a_sign.c:224:

```

Than the issue was introduced between v2.2.3 and v2.3.0.

**Comment by AndreasFuchsSIT** [comment]
If 2.2.0 works and 2.3.0 does not, you could go ahead and use
git bisect to detect the commit along the master series that caused the issue.

________________________________

**Comment by nesvera** [comment]
It's been quite some time that I found a solution for it, but did not had time to update it here.

I follow the suggestions to bisect between the version and noticed that the problem was gone once it was necessary to downgrade the tpm2-tss-engine. Later, I discovered that the issue was introduced in tpm2-tss-engine between version 1.0.1 and 1.1.0. Nowadays, I am using the version 1.0.1 that is working great.

I was not able to find the exact commit that introduced the issue on tpm2-tss-engine because at some point I had the following error during compile time: **undefined reference to `Tss2_MU_TPM2B_PUBLIC_Unmarshal'** .

The undefined error started on commit 760ab9fae11b8011655a09c82af06c9e86c71e7e 
The initial error (0x000001df) started on commit 6c31844e86bbeb56c6c5d78e91e703e7cae26ab3 

There is a remarkable thing that I found quite strange. I have two hardware versions from Infeneon's TPM2.0: SLB9670 and SLB9665. The first presents the error 0x000001df and the later not. I did had time to investigate the diferences between each version.

That being said we can close this issue once it is not related to this repo. Should I open an issue on tpm2-tss-engine's repo to track the possible causes?

Thanks.
